File: P357_PrimeGeneratingIntegers.py
# NOTE: Need more optimization. Takes a few minutes for 10^7 without using sieve of precalculated primes. Maybe try writing it in C.

# 524402304 till 10^6
# 27814470276 till 10^7

import math

def isPrime(n):
    prime = True
    if n <= 1:
        return False
    elif n == 2:
        return True
    elif n % 2 == 0:
        return False
    else:
        i = 3
        limit = n ** 0.5
        while i <= limit:
            if n % i == 0:
                return False
            i += 2
        return True

sum = 0


def SieveOfEratosthenesArray(limit):
    primes = [True] * limit
    primeNumbers = []
    primes[0] = primes[1] = False
    for i, isPrime in enumerate(primes):
        if isPrime:
            primeNumbers.append(i)
            for n in range(i, limit, i):
                primes[n] = False
    return primeNumbers


def satisfies(n):
    i = 2
    while i <= math.sqrt(n):
        if (n % i == 0):
            if (n / i == i):
                return False
            elif i + n / i not in primes:
                return False
        i = i + 1
    return True

limit = 10 ** 7

primes = SieveOfEratosthenesArray(limit)

for i in range(2, limit, 4):
    if isPrime(i + 1) and isPrime(2 + i / 2) and satisfies(i):
        sum += i

print(sum)

File: P357_PrimeGeneratingIntegers.js
/*
Consider the divisors of 30: 1,2,3,5,6,10,15,30.
It can be seen that for every divisor d of 30, d+30/d is prime.

Find the sum of all positive integers n not exceeding 100 000 000
such that for every divisor d of n, d+n/d is prime.

Answer: 1739023853137 (99s)
*/

const sieve = require('./Useful Programs/SieveOfEratosthenes');
const getDivisors = require('./Useful Programs/AllFactors').getProperDivisors;
const isPrime = require('./Useful Programs/PrimeTest');

const limit = 10 ** 8; 
const primes = sieve(limit);

const satisfies = n => {
    const factors = getDivisors(n);
    const halfFactors = factors.slice(0, Math.ceil(factors.length / 2));
    for (const factor of halfFactors.slice(2,)) {
        // if (!binarySearch(primes, (factor + n / factor))) {
        //     return false;
        // }
        if (!isPrime((factor + n / factor))) {
            return false;
        }
    }
    return true;
};

let sum = 1; // The number '1' satisfies the conditions but I don't include it in the loop below so I'm starting the sum from '1'. 

// All such numbers have to be of the form 4k+2 for all k=0,1,2,...etc.
for (let i = 2; i <= limit; i += 4) {
    if (isPrime(i + 1) && isPrime(2 + i / 2) && satisfies(i)) {
        sum += i;
    }
}

console.log(sum);

//27814470276 till 10^7 in 4.5s
// 1739023853136 till 10^8 in 99s. (Answer is 1 more than this).